linera-client/src/client_context.rs

```diff
@@ -566,16 +567,55 @@ where
     W: Persist<Target = Wallet>,
 {
     pub async fn process_inboxes_and_force_validator_updates(&mut self) {
-        for chain_id in self.wallet.owned_chain_ids() {
-            let chain_client = self
-                .make_chain_client(chain_id)
-                .expect("chains in the wallet must exist");
-            self.process_inbox(&chain_client).await.unwrap();
-            chain_client.update_validators(None).await.unwrap();
-            self.update_wallet_from_client(&chain_client).await.unwrap();
+        let chain_clients = self
+            .wallet
+            .owned_chain_ids()
+            .iter()
+            .map(|chain_id| {
+                self.make_chain_client(*chain_id)
+                    .expect("chains in the wallet must exist")
+            })
+            .collect::<Vec<_>>();
+
+        let mut join_set = task::JoinSet::new();
+        for chain_client in chain_clients {
+            join_set.spawn(async move {
+                Self::process_inbox_without_updating_wallet(&chain_client)
+                    .await
+                    .expect("Processing inbox should not fail!");
+                chain_client.update_validators(None).await.unwrap();
+                chain_client
+            });
+        }
+
+        let chain_clients = join_set.join_all().await;
+        for chain_client in &chain_clients {
+            self.update_wallet_from_client(chain_client).await.unwrap();
+        }
+        if chain_clients.is_empty() {
+            self.save_wallet().await.unwrap();
         }
     }
 
+    async fn process_inbox_without_updating_wallet(
+        chain_client: &ChainClient<NodeProvider, S>,
+    ) -> Result<Vec<ConfirmedBlockCertificate>, Error> {
+        let mut certificates = Vec::new();
+        // Try processing the inbox optimistically without waiting for validator notifications.
+        let (new_certificates, maybe_timeout) = {
+            chain_client.synchronize_from_validators().await?;
+            let result = chain_client.process_inbox_without_prepare().await;
+            result?
+        };
```

Comment:
Why the curly brackets?

```suggestion
        chain_client.synchronize_from_validators().await?;
        let (new_certificates, maybe_timeout) = chain_client.process_inbox_without_prepare().await?;
```

Reply:
True, I guess now it's not necessary anymore. Thanks!